File: erlang/euler2.erl
-module(euler2).
-export[euler2/0].

% Sum of all even valued fibonacci numbers below 2.000.000
euler2() ->
    lists:foldl(
        fun(X, Sum) -> X + Sum end, 
        0, 
	[X || X <- fibsbelow(2000000), X rem 2 =:= 0]).

fibsbelow(1) -> [];
fibsbelow(2) -> [1];
fibsbelow(N) -> fibsbelow(N, [2, 1]).
fibsbelow(N, Fibs) when hd(Fibs) >= N -> lists:reverse(tl(Fibs));
fibsbelow(N, Fibs) -> fibsbelow(N, [hd(Fibs) + hd(tl(Fibs)) | Fibs]).

File: clojure/euler2.clj
; Calculates sum of even-valued fibonacci numbers below 4.000.000
(defn fibs 
    ([a b] (cons a (lazy-seq (fibs b (+ a b))))))

(println (
    reduce
        +
	(filter 
            (fn [x] (= (mod x 2) 0))
	    (take-while (fn [x] (< x 4000000)) (fibs 1 2)))))

